h/api/models.py

```diff
@@ -11,9 +11,45 @@
 from annotator import document
 
 
-def is_file(uri):
-    """Return True if the given uri is a local file uri."""
-    return uri and uri.lower().startswith("file:///")
+def _format_document_link(href, title, link_text, hostname):
+    """Return a document link for the given components.
+
+    Helper function for the .document_link property below.
+
+    The given href, title, link_text and hostname are assumed to be already
+    safely escaped. The returned string will be a Markup object so that
+    it can be rendered in Jinja2 templates without further escaped occurring.
+
+    """
+    if hostname and hostname in link_text:
+        hostname = ""
+
+    def truncate(content, length=50):
+        """Truncate the given string to at most length chars."""
+        if len(content) <= length:
+            return content
+        else:
+            return content[:length] + jinja2.Markup("&hellip;")
+
+    hostname = truncate(hostname)
+    link_text = truncate(link_text)
+
+    if href and hostname:
+        link = ('<a href="{href}" title="{title}">{link_text}</a> '
+                '({hostname})'.format(href=href, title=title,
+                                      link_text=link_text, hostname=hostname))
+    elif hostname and not href:
+        link = ('<a title="{title}">{link_text}</a> ({hostname})'.format(
+            title=title, link_text=link_text, hostname=hostname))
+    elif href and not hostname:
+        link = '<a href="{href}" title="{title}">{link_text}</a>'.format(
+            href=href, title=title, link_text=link_text)
+    else:
+        link = '<a title="{title}">{link_text}</a>'.format(
+            title=title, link_text=link_text)
+
```

Comment:
This whole function would be a fair bit clearer if you just assigned the format strings and did the `.format` once at the end:

```
if href and hostname:
    tpl = '<a href="{href}" title="{title}">{link_text}</a> ({hostname})'
elif hostname:
    tpl = '...'
elif href:
   ...

link = tpl.format(href=href, title=title, link_text=link_text, hostname=hostname)
return jinja2.Markup(link)
```